crates/bevy_pbr/src/ssr/mod.rs

```diff
@@ -0,0 +1,563 @@
+    ) in &views
+    {
+        // SSR is only supported in the deferred pipeline, which has no MSAA
+        // support. Thus we can assume MSAA is off.
+        let mut mesh_pipeline_view_key = MeshPipelineViewLayoutKey::from(Msaa::Off)
+            | MeshPipelineViewLayoutKey::DEPTH_PREPASS
+            | MeshPipelineViewLayoutKey::DEFERRED_PREPASS;
+        mesh_pipeline_view_key.set(
+            MeshPipelineViewLayoutKey::NORMAL_PREPASS,
+            has_normal_prepass,
+        );
+        mesh_pipeline_view_key.set(
+            MeshPipelineViewLayoutKey::MOTION_VECTOR_PREPASS,
+            has_motion_vector_prepass,
+        );
+
+        // Build the pipeline.
+        let pipeline_id = pipelines.specialize(
+            &pipeline_cache,
+            &ssr_pipeline,
+            ScreenSpaceReflectionsPipelineKey {
+                mesh_pipeline_view_key,
+                is_hdr: extracted_view.hdr,
+                has_environment_maps,
+            },
+        );
+
+        // Note which pipeline ID was used.
+        commands
+            .entity(entity)
+            .insert(ScreenSpaceReflectionsPipelineId(pipeline_id));
+    }
+}
+
+/// Gathers up screen space reflection settings for each applicable view and
+/// writes them into a GPU buffer.
+pub fn prepare_ssr_settings(
+    mut commands: Commands,
+    views: Query<(Entity, Option<&ScreenSpaceReflectionsUniform>), With<ExtractedView>>,
+    mut ssr_settings_buffer: ResMut<ScreenSpaceReflectionsBuffer>,
+    render_device: Res<RenderDevice>,
+    render_queue: Res<RenderQueue>,
+) {
+    let Some(mut writer) =
+        ssr_settings_buffer.get_writer(views.iter().len(), &render_device, &render_queue)
+    else {
+        return;
+    };
+
+    for (view, ssr_uniform) in views.iter() {
+        let uniform_offset = match ssr_uniform {
+            None => 0,
+            Some(ssr_uniform) => writer.write(ssr_uniform),
+        };
+        commands
+            .entity(view)
+            .insert(ViewScreenSpaceReflectionsUniformOffset(uniform_offset));
+    }
+}
+
+impl ExtractComponent for ScreenSpaceReflectionsSettings {
+    type QueryData = Read<ScreenSpaceReflectionsSettings>;
+
+    type QueryFilter = ();
+
+    type Out = ScreenSpaceReflectionsUniform;
+
+    fn extract_component(settings: QueryItem<'_, Self::QueryData>) -> Option<Self::Out> {
+        if !DEPTH_TEXTURE_SAMPLING_SUPPORTED {
+            info_once!(
+                "Disabling screen-space reflections on this platform because depth textures \
+                aren't supported correctly"
+            );
+            return None;
+        }
+
+        Some((*settings).into())
+    }
+}
+
+impl SpecializedRenderPipeline for ScreenSpaceReflectionsPipeline {
+    type Key = ScreenSpaceReflectionsPipelineKey;
+
+    fn specialize(&self, key: Self::Key) -> RenderPipelineDescriptor {
+        let mesh_view_layout = self
+            .mesh_view_layouts
+            .get_view_layout(key.mesh_pipeline_view_key);
```

Comment:
```suggestion
        let mesh_view_layout = self.get_view_layout(key.mesh_pipeline_view_key);
```

Reply:
Doesn't compile. `self` is `ScreenSpaceReflectionsPipeline`, not `MeshPipeline`.

Reply:
oh, right, sorry, that's mesh_view_layouts not view_layouts 😅